### ISSUE TITLE
Can't import ResNet50 from Keras(th) on master

### ISSUE THREAD

**Issue by rohrl** [deeplearning4j/deeplearning4j#2901]
I got around the ZeroPadding2D by removing it and adding "border_mode='same'", but further on DL4J import throws:
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.CommandLineWrapper.main(CommandLineWrapper.java:130)
Caused by: java.lang.IllegalStateException: Mis matched lengths: [32768] != [16384]
	at org.nd4j.linalg.util.LinAlgExceptions.assertSameLength(LinAlgExceptions.java:40)
	at org.nd4j.linalg.api.ops.BaseTransformOp.<init>(BaseTransformOp.java:48)
	at org.nd4j.linalg.api.ops.impl.transforms.Set.<init>(Set.java:25)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.assign(BaseNDArray.java:1203)
	at org.deeplearning4j.nn.layers.BaseLayer.setParam(BaseLayer.java:281)
	at org.deeplearning4j.nn.modelimport.keras.KerasLayer.copyWeightsToLayer(KerasLayer.java:432)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.helperCopyWeightsToModel(KerasModel.java:598)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:453)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:439)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasModelAndWeights(KerasModelImport.java:205)
	at com.nuix.load.image.ModelImport.main(ModelImport.java:43)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
	... 5 more
```
Here is the json config from Keras:
https://gist.github.com/rohrl/a456e8cb775b6f7d49c7ec0d45db5f05
I'm using this together with the weights file generated using `model.save_weights`

**Comment by turambar** [comment]
Note: this (just merged) PR adds explicit support for ZeroPadding layers: https://github.com/deeplearning4j/deeplearning4j/pull/2849

**Comment by turambar** [comment]
This was a bug in import of functional API models, and has been fixed in this merged PR: https://github.com/deeplearning4j/deeplearning4j/pull/2903. I can confirm I successfully imported the model, including weights, from your gist. Closing.

**Comment by rohrl** [comment]
Hi, I'm still seeing errors with ResNet50.

I used CPU build from master at rev fbac773 (so including your PR)

The network I'm importing is ResNet50 from Keras with theano backend:
https://github.com/fchollet/keras/blob/master/keras/applications/resnet50.py
https://github.com/fchollet/deep-learning-models/releases/download/v0.2/resnet50_weights_th_dim_ordering_th_kernels.h5

When I use `KerasModelImport.importKerasModelAndWeights(String jsonConfig, String hd5fileWeights)` I get the same error as previously:
`java.lang.IllegalStateException: Mis matched lengths: [32768] != [16384]`

I also tried using `KerasModelImport.importKerasModelAndWeights(String hd5fileName)` - this fails with following:
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.CommandLineWrapper.main(CommandLineWrapper.java:130)
Caused by: java.lang.RuntimeException: Unknown exception.
	at org.bytedeco.javacpp.hdf5$H5Location.openAttribute(Native Method)
	at org.deeplearning4j.nn.modelimport.keras.Hdf5Archive.readAttributeAsJson(Hdf5Archive.java:88)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel$ModelBuilder.modelHdf5Filename(KerasModel.java:525)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasModelAndWeights(KerasModelImport.java:122)
	at com.nuix.tools.deeplearning.ModelImport.main(ModelImport.java:69)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
	... 5 more
HDF5-DIAG: Error detected in HDF5 (1.10.0-patch1) thread 0:
  #000: C:\autotest\HDF5110ReleaseRWDITAR\src\H5A.c line 438 in H5Aopen(): unable to load attribute info from object header for attribute: 'model_config'
    major: Attribute
    minor: Unable to initialize object
  #001: C:\autotest\HDF5110ReleaseRWDITAR\src\H5Oattribute.c line 530 in H5O_attr_open_by_name(): can't locate attribute: 'model_config'
    major: Attribute
    minor: Object not found
```

**Comment by turambar** [comment]
@rohrl I'm giving it a shot now.

**Comment by turambar** [comment]
Reopening as I've been able to reproduce this issue. That said, I'm not 100% convinced this is on our end -- either that, or I'm misunderstanding something about this model. As far as I can tell, this Exception is being thrown when we try to set the weights for the `res2b_branch2a` convolutional layer. As far as I can tell, the input to that layer has shape (h=55, w=55, d=512), but we're trying to set weights with shape (h=1, w=1, channels=256, filters=64), which makes very little sense.

@rohrl can you offer any insights about this?

**Comment by rohrl** [comment]
Sadly, I don't know anything more.
I've been trying to import net with weights from Keras and getting this
error.
Is there any other source of trained Resnet ? You said you managed to
import Resnet with weights successfully - which model was it?
Will TF backend work ?

**Comment by turambar** [comment]
@rohrl OK, I figured it out. Keras [merge layers](https://keras.io/layers/core/#merge) have a `mode` field that determines how to merge the layers, whether via concatenation or summation, etc. Currently, we've been directly mapping Keras merge layers to DL4J Merge vertices, but when the mode is something other than concat (e.g., sum), we should map it to a DL4J ElementWiseVertex. Will fix before next release and post an update here once done.

**Comment by rohrl** [comment]
Thanks @turambar . So is there any possibility at the moment to import ResNet with weights to DL4J ?

**Comment by turambar** [comment]
Fixed in this merged PR: https://github.com/deeplearning4j/deeplearning4j/pull/2915

NOTE: can only confirm successful import without Exception. Have not tested performance.

**Comment by rohrl** [comment]
@turambar Hey, thanks, I confirm that import was successful.
However when I run it, it spits this out:
```
Caused by: org.nd4j.shade.jackson.databind.JsonMappingException: No suitable constructor found for type [simple type, class org.deeplearning4j.nn.conf.layers.ZeroPaddingLayer]: can not instantiate from JSON object (missing default constructor or creator, or perhaps need to add/enable type information?)
 at [Source: java.io.StringReader@5ce675b5; line: 220, column: 15] (through reference chain: org.deeplearning4j.nn.conf.ComputationGraphConfiguration["vertices"]->java.util.LinkedHashMap["zeropadding2d_1"]->org.deeplearning4j.nn.conf.graph.LayerVertex["layerConf"]->org.deeplearning4j.nn.conf.NeuralNetConfiguration["layer"])
	at org.nd4j.shade.jackson.databind.JsonMappingException.from(JsonMappingException.java:148)
	at org.nd4j.shade.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1080)
	at org.nd4j.shade.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:295)
	at org.nd4j.shade.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:142)
	at org.nd4j.shade.jackson.databind.jsontype.impl.AsWrapperTypeDeserializer._deserialize(AsWrapperTypeDeserializer.java:114)
	at org.nd4j.shade.jackson.databind.jsontype.impl.AsWrapperTypeDeserializer.deserializeTypedFromObject(AsWrapperTypeDeserializer.java:49)
	at org.nd4j.shade.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:131)
	at org.nd4j.shade.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:521)
	at org.nd4j.shade.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:95)
	at org.nd4j.shade.jackson.databind.deser.impl.BeanPropertyMap.findDeserializeAndSet(BeanPropertyMap.java:285)
	at org.nd4j.shade.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:248)
	at org.nd4j.shade.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:136)
	at org.nd4j.shade.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:523)
	at org.nd4j.shade.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:95)
	at org.nd4j.shade.jackson.databind.deser.impl.BeanPropertyMap.findDeserializeAndSet(BeanPropertyMap.java:285)
	at org.nd4j.shade.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:248)
	at org.nd4j.shade.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:136)
	at org.nd4j.shade.jackson.databind.jsontype.impl.AsWrapperTypeDeserializer._deserialize(AsWrapperTypeDeserializer.java:114)
	at org.nd4j.shade.jackson.databind.jsontype.impl.AsWrapperTypeDeserializer.deserializeTypedFromObject(AsWrapperTypeDeserializer.java:49)
	at org.nd4j.shade.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:131)
	at org.nd4j.shade.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:475)
	at org.nd4j.shade.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:330)
	at org.nd4j.shade.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:25)
	at org.nd4j.shade.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:523)
	at org.nd4j.shade.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:95)
	at org.nd4j.shade.jackson.databind.deser.impl.BeanPropertyMap.findDeserializeAndSet(BeanPropertyMap.java:285)
	at org.nd4j.shade.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:248)
	at org.nd4j.shade.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:136)
	at org.nd4j.shade.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3562)
	at org.nd4j.shade.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2578)
	at org.deeplearning4j.nn.conf.ComputationGraphConfiguration.fromJson(ComputationGraphConfiguration.java:137)
	... 18 more
```

**Comment by rohrl** [comment]
I just run it and the imported network seems to be broken - it always outputs similar values ("cleaver; dishwasher; nail; nematode; oboe") regardless of input.

**Comment by turambar** [comment]
(98% copy and paste of my comment on the Inception issue!)

@rohrl thank you for your persistence in testing model import. I'm traveling this week so I don't have a ton of dev cycles, but I'll try to take a look when I settle back down (weekend or early next week at latest). In the meantime, I encourage you to fire up your debugger and dive into the source code -- even if you just narrow it down some, that will help accelerate a fix, if one is needed.

@eraly if you get any bandwidth, maybe take a look? This is also on our Keras model import post-release test TODO list anyway, no?

**Comment by turambar** [comment]
Will take a look at this this week.

**Comment by agibsonccc** [comment]
Was this fixed?

**Comment by turambar** [comment]
Not yet.

**Comment by maxpumperla** [comment]
@rohrl what's the status of this issue? You can import now, but output is off / inconsistent? Thanks

**Comment by maxpumperla** [comment]
note to self: the batchnorm axis mostly used for this resnet model (`axis=2`) is currently not supported.

**Comment by raver119** [comment]
@maxpumperla are you sure new batchnorm impl doesn't support it?

**Comment by dpratt** [comment]
I'm seeing a very similar problem - I have a model trained up with the stock Keras 2.0 ResNet50 implementation, and attemping to load it gives the following using 2.0.9-SNAPSHOT (from today) of dl4j.

```
java.lang.IllegalStateException: Mis matched lengths: [32768] != [16384] - Array 1 shape: [64, 512, 1, 1], array 2 shape: [64, 256, 1, 1]
	at org.nd4j.linalg.util.LinAlgExceptions.assertSameLength(LinAlgExceptions.java:41)
	at org.nd4j.linalg.api.ops.BaseTransformOp.<init>(BaseTransformOp.java:174)
	at org.nd4j.linalg.api.ops.impl.transforms.Set.<init>(Set.java:41)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.assign(BaseNDArray.java:1273)
	at org.deeplearning4j.nn.layers.BaseLayer.setParam(BaseLayer.java:204)
	at org.deeplearning4j.nn.modelimport.keras.KerasLayer.copyWeightsToLayer(KerasLayer.java:295)
	at org.deeplearning4j.nn.modelimport.keras.utils.KerasModelUtils.copyWeightsToModel(KerasModelUtils.java:76)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:377)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:363)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasModelAndWeights(KerasModelImport.java:107)
```
